### PR TITLE
Fix #1

### LINKED ISSUE #1: Inclusion request for Conan Center
## Inclusion Request Review

This is a review for [Exiv2/0.27@piponazo/stable(piponazo/piponazo)](https://github.com/piponazo/conan-exiv2) inclusion request to conan-center.
This review follows the [inclusion guidelines of third party libraries](http://docs.conan.io/en/latest/uploading_packages/bintray/conan_center_guide.html#inclusion-guidelines-for-third-party-libraries).

### Package
Exiv2/0.27@piponazo/stable

#### Recipe quality
- [x] Git public repository
- [x] Recipe fields
- [x] Linter
- [x] Updated:
  - Please update the recipe to follow the latest recipe conventions as announced [here](https://blog.conan.io/2018/10/11/New-conan-release-1-8.html)
  - Use ``cmake.defintions`` dict to set any CMake definition instead of ``cmake_args``
- [x] Clean
- [x] test_package
- [x] Raise errors on invalid configurations
- [x] LICENSE of the recipe
- [x] LICENSE of the library: Please package the license of the library in the ``package()`` method

#### CI Integration
- [x] Provide at least one CI configuration (header only)
- [x] (If not header only) CI configurations for Linux/MacOSX/Windows

#### Bintray Package Information
Please update the following Bintray information (unchecked ones).
- [x] Description
- [x] Licenses
- [x] Tags
- [x] Maturity
- [x] Website (if any)
- [x] Issue tracker
- [x] Version control

### Additional comments

- There are some outdated packages, clean them up with: 
`conan remove Exiv2/0.27@piponazo/testing --outdated -r bincrafters

- There are testing packages that shouldn't be uploaded to Bintray. Please delete them in order to accept the inclusion request.

- There are missing packages for Visual Studio 12, 14 and 15

### PR DESCRIPTION
**Pull request by piponazo**
@danimtb this should address all the comments you made in #1.

I have some questions:
* What does it mean the "clean" checkbox?
* You said that "There are missing packages for Visual Studio 12, 14 and 15" but the only missing ones were the ones for Visual Studio 12. Unless I am missing something :sweat_smile: 

Please, let me know if you see something which could be improved.

### CODE REVIEW

File: conanfile.py [hunk 1]
```diff
@@ -75,3 +76,6 @@ def package_info(self):
         self.cpp_info.debug.libs = ["exiv2d"]
         self.cpp_info.release.libs = ["exiv2"]
         self.cpp_info.libdirs = ['lib']  # Directories where libraries can be found
```
Comment (danimtb) [inline conanfile.py]: FYI: This is the default value of `includedirs`

Comment (danimtb) [inline conanfile.py]: FYI: This is the default value of `libdirs`

File: conanfile.py [hunk 2]
```diff
@@ -23,50 +23,51 @@ class Exiv2Conan(ConanFile):
...
+        cmake.definitions["CMAKE_INSTALL_PREFIX"] = self.package_folder
 
         if tools.os_info.is_windows:
-            cmake_args['EXIV2_ENABLE_WIN_UNICODE'] = self.options.unicode
+            cmake.definitions['EXIV2_ENABLE_WIN_UNICODE'] = self.options.unicode
 
-        cmake.configure(source_dir="../exiv2", build_dir="build", defs=cmake_args)
+        cmake.configure(source_dir="../exiv2", build_dir="build")
         cmake.build()
         cmake.install()
```
Comment (danimtb) [inline conanfile.py:71]: FYI: Some people do the install in the package method (cmake.deifntions + cmake.configure() + cmake.install) to be able to use package development commands independently (conan build + conan package).

File: conanfile.py [hunk 3]
```diff
@@ -23,50 +23,51 @@ class Exiv2Conan(ConanFile):
...
+        cmake.definitions["EXIV2_ENABLE_VIDEO"] = self.options.video
+        cmake.definitions["EXIV2_ENABLE_WEBREADY"] = self.options.webready
+        cmake.definitions["EXIV2_ENABLE_CURL"] = self.options.curl
+        cmake.definitions["EXIV2_ENABLE_SSH"] = self.options.ssh
+        cmake.definitions["EXIV2_BUILD_SAMPLES"] = "OFF"
+        cmake.definitions["EXIV2_BUILD_EXIV2_COMMAND"] = self.options.buildTool
+        cmake.definitions["EXIV2_BUILD_PO"] = "OFF"
+        cmake.definitions["EXIV2_ENABLE_XMP"] = self.options.xmp
+        cmake.definitions["EXIV2_ENABLE_PNG"] = self.options.png
+        cmake.definitions["CMAKE_INSTALL_PREFIX"] = self.package_folder
```
Comment (danimtb) [inline conanfile.py]: This is already done by the helper, see https://docs.conan.io/en/latest/reference/build_helpers/cmake.html#definitions

File: conanfile.py [hunk 4]
```diff
@@ -23,50 +23,51 @@ class Exiv2Conan(ConanFile):
...
 
         if tools.os_info.is_windows:
-            cmake_args['EXIV2_ENABLE_WIN_UNICODE'] = self.options.unicode
+            cmake.definitions['EXIV2_ENABLE_WIN_UNICODE'] = self.options.unicode
 
-        cmake.configure(source_dir="../exiv2", build_dir="build", defs=cmake_args)
+        cmake.configure(source_dir="../exiv2", build_dir="build")
```
Comment (danimtb) [inline conanfile.py]: Use `source_folder` and `build_folder` instead. These parameters are deprecated https://docs.conan.io/en/latest/reference/build_helpers/cmake.html#configure

**Review by danimtb — CHANGES_REQUESTED** [review]
Thanks a lot for the update! Did some minor comments 😄

**Comment by danimtb** [comment]
Sorry, I forgot to tick the "clean" checkbox. It usually means that the recipe does too many complex things or the code is nor very readable/maintainable (not your case).

Seems there are some missing packages of  Expat/2.2.6@pix4d/stable for clang 7, you would need to add those in its Travis matrix. There was also a random connection issue with Bintray, just realunch the job.

Please, take a look at the test_package recipe too. It will need an update on the cmake helper (source_dir/build_dir) and defining the user and channel as well as the requirement is no longer needed https://github.com/piponazo/conan-exiv2/blob/5cbfe04df7842264cafd7662fc234161775b13aa/test_package/conanfile.py#L9

Another final comment is the build type of this package. Currently it seems that the package builds by default both Release and Debug versions, so there is no need to create a package for each of them. However, I think the setting needs to be there in case someone wants the build type as "MinSizeRel" or "RelWithDebInfo". I this case you could tune up the package_id with the following:

```
def package_id(self):
    if self.settings.build_type == "Release" or self.settings.build_type == "Debug":
        self.info.settings.build_type = "DebugRelease"  # The value doesn't matter, only being the same for Debug and Release
```

You can find more information here https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html#defining-a-custom-package-id

Hope it makes sense and thanks a lot for addressing the points in the review 😄

**Comment by piponazo** [comment]
> Seems there are some missing packages of Expat/2.2.6@pix4d/stable for clang 7

Yes, I noticed. I am also updating that recipe 😉 . I will try to apply the same improvements there that you commented here.

> Another final comment is the build type of this package. Currently it seems that the package builds by default both Release and Debug versions, so there is no need to create a package for each of them.

😲 I did not know about that `DebugRelease` thing ... I will double check. 

> Hope it makes sense and thanks a lot for addressing the points in the review 😄

Thanks to you for the detailed review. I am glad to learn to write better recipes 👍

File: conanfile.py (hunk 2 above)
Comment (piponazo) [reply conanfile.py:71]: I read about this few months ago and It's good to know, however I never had the need to go so far. I will not implement this change for the moment.

**Comment by piponazo** [comment]
Yesterday I implemented most of the changes you proposed. Just to comments more:

> Please, take a look at the test_package recipe too. It will need an update on the cmake helper (source_dir/build_dir) and defining the user and channel as well as the requirement is no longer needed

It seems that you were looking to the version in master with is an old version of the recipe. Those things are fixed in this branch. 

> Currently it seems that the package builds by default both Release and Debug versions ...

I understood later what you meant with this, and it is not the case. We only compile one version of the library depending on the `CMAKE_BUILD_TYPE` value.

Now all the CI jobs look fine, although not the ones you can see in the CI checks in this PR since I have been triggering some jobs manually.

**Comment by danimtb** [comment]
Sorry, my fault. I looked at the master branch as you figured out.

Regarding the build type, I thought it was building both release an debug types regarding the package_info https://github.com/piponazo/conan-exiv2/blob/a6992125462053637e81411019db0a2a56bb13e8/conanfile.py#L73 but seems it is not the case 🙃

**Pull request merged**